File: src/main/kotlin/com/feakin/intellij/structure/FeakinStructureViewModel.java
package com.feakin.intellij.structure;

import com.feakin.intellij.FeakinFile;
import com.intellij.ide.structureView.StructureViewModel;
import com.intellij.ide.structureView.StructureViewModelBase;
import com.intellij.ide.structureView.StructureViewTreeElement;
import com.intellij.ide.util.treeView.smartTree.Sorter;
import com.intellij.psi.PsiFile;
import org.jetbrains.annotations.NotNull;

public class FeakinStructureViewModel extends StructureViewModelBase implements
        StructureViewModel.ElementInfoProvider {

    public FeakinStructureViewModel(PsiFile psiFile) {
        super(psiFile, new FeakinStructureViewElement(psiFile));
    }

    @NotNull
    public Sorter[] getSorters() {
        return new Sorter[]{Sorter.ALPHA_SORTER};
    }


    @Override
    public boolean isAlwaysShowsPlus(StructureViewTreeElement element) {
        return false;
    }

    @Override
    public boolean isAlwaysLeaf(StructureViewTreeElement element) {
        return element instanceof FeakinFile;
    }
}

File: src/main/kotlin/com/feakin/intellij/structure/FeakinStructureViewModel.kt
package com.feakin.intellij.structure

import com.feakin.intellij.FkFile
import com.feakin.intellij.psi.FeakinAggregateDeclaration
import com.feakin.intellij.psi.FeakinContextDeclaration
import com.feakin.intellij.psi.FeakinContextMapDeclaration
import com.intellij.ide.structureView.StructureViewModel.ElementInfoProvider
import com.intellij.ide.structureView.StructureViewModelBase
import com.intellij.ide.structureView.StructureViewTreeElement
import com.intellij.ide.util.treeView.smartTree.Sorter
import com.intellij.psi.PsiFile

class FeakinStructureViewModel(psiFile: PsiFile) : StructureViewModelBase(psiFile, FeakinStructureViewElement(psiFile)),
    ElementInfoProvider {

    override fun getSorters(): Array<Sorter> = arrayOf(Sorter.ALPHA_SORTER)

    override fun isAlwaysShowsPlus(element: StructureViewTreeElement): Boolean = element.value is FkFile

    override fun isAlwaysLeaf(element: StructureViewTreeElement): Boolean {
        return when (element.value) {
            is FeakinContextMapDeclaration,
            is FeakinContextDeclaration,
            is FeakinAggregateDeclaration,
            -> true

            else -> false
        }
    }
}